File: frontend/src/lib/api/query-analytics.ts
import { apiRequest } from "@/lib/api/request";

// ── Shared types ───────────────────────────────────────────────────────────────

export type QueryAnalyticsDateRange = {
  from_date: string;
  to_date: string;
};

// ── Summary ────────────────────────────────────────────────────────────────────

export type FeedbackCategoryCount = {
  category: string;
  count: number;
};

export type QueryAnalyticsSummaryResponse = {
  organization_id: string;
  range: QueryAnalyticsDateRange;
  generated_at: string;
  enabled: boolean;
  disabled_reason: string | null;
  total_queries: number;
  answered_queries: number;
  unanswered_queries: number;
  low_confidence_queries: number;
  negative_feedback_count: number;
  unanswered_rate: number | null;
  avg_confidence: number | null;
  negative_feedback_rate: number | null;
  top_feedback_categories: FeedbackCategoryCount[];
  top_feedback_reasons: FeedbackCategoryCount[];
};

export type QueryAnalyticsSummaryQuery = {
  from?: string;
  to?: string;
};

export async function getQueryAnalyticsSummary(
  query: QueryAnalyticsSummaryQuery = {},
): Promise<QueryAnalyticsSummaryResponse> {
  return apiRequest<QueryAnalyticsSummaryResponse>(
    "/admin/query-analytics/summary",
    { query },
  );
}

// ── Trends ─────────────────────────────────────────────────────────────────────

export type QueryTrendPoint = {
  date: string;
  total_queries: number;
  unanswered: number;
  low_confidence: number;
  negative_feedback: number;
  avg_confidence: number | null;
};

export type QueryTrendsResponse = {
  organization_id: string;
  range: QueryAnalyticsDateRange;
  generated_at: string;
  points: QueryTrendPoint[];
};

export async function getQueryAnalyticsTrends(
  query: QueryAnalyticsSummaryQuery = {},
): Promise<QueryTrendsResponse> {
  return apiRequest<QueryTrendsResponse>("/admin/query-analytics/trends", {
    query,
  });
}

// ── Knowledge Gaps ─────────────────────────────────────────────────────────────

export type GapType =
  | "no_answer"
  | "low_confidence"
  | "bad_feedback"
  | "stale_citation"
  | "missing_source";

export type GapStatus = "open" | "in_review" | "resolved" | "dismissed";

export type GapSource =
  | "admin"
  | "low_confidence_analysis"
  | "feedback_analysis"
  | "no_answer_analysis";

export type KnowledgeGapResponse = {
  gap_id: string;
  organization_id: string;
  gap_type: GapType;
  topic_label: string;
  description: string | null;
  gap_source: GapSource;
  occurrence_count: number;
  avg_confidence: number | null;
  example_query: string | null;
  status: GapStatus;
  remediation_json: Record<string, unknown> | null;
  collection_id: string | null;
  linked_document_id: string | null;
  linked_eval_question_id: string | null;
  converted_to: "eval_case" | "doc_request" | "review_task" | null;
  converted_at: string | null;
  reviewer_notes: string | null;
  created_at: string;
  updated_at: string;
};

export type KnowledgeGapListResponse = {
  items: KnowledgeGapResponse[];
  total: number;
};

export type ListGapsOptions = {
  status?: GapStatus;
  gap_type?: GapType;
  limit?: number;
  offset?: number;
};

export async function listKnowledgeGaps(
  options: ListGapsOptions = {},
): Promise<KnowledgeGapListResponse> {
  return apiRequest<KnowledgeGapListResponse>("/admin/query-analytics/gaps", {
    query: {
      status: options.status,
      gap_type: options.gap_type,
      limit: options.limit,
      offset: options.offset,
    },
  });
}

export type CreateKnowledgeGapRequest = {
  gap_type: GapType;
  topic_label: string;
  description?: string | null;
  occurrence_count?: number;
  avg_confidence?: number | null;
  example_query?: string | null;
  collection_id?: string | null;
  gap_source?: GapSource;
};

export async function createKnowledgeGap(
  payload: CreateKnowledgeGapRequest,
): Promise<KnowledgeGapResponse> {
  return apiRequest<KnowledgeGapResponse>("/admin/query-analytics/gaps", {
    method: "POST",
    json: payload,
  });
}

export type UpdateKnowledgeGapRequest = {
  status?: GapStatus;
  reviewer_notes?: string | null;
  linked_document_id?: string | null;
  description?: string | null;
};

export async function updateKnowledgeGap(
  gapId: string,
  payload: UpdateKnowledgeGapRequest,
): Promise<KnowledgeGapResponse> {
  return apiRequest<KnowledgeGapResponse>(
    `/admin/query-analytics/gaps/${encodeURIComponent(gapId)}`,
    { method: "PATCH", json: payload },
  );
}

export type ConvertKnowledgeGapRequest = {
  target: "eval_case" | "doc_request" | "review_task";
  evaluation_set_id?: string | null;
  notes?: string | null;
};

export type ConvertKnowledgeGapResponse = {
  gap_id: string;
  converted_to: string;
  converted_at: string;
  linked_eval_question_id: string | null;
};

export async function convertKnowledgeGap(
  gapId: string,
  payload: ConvertKnowledgeGapRequest,
): Promise<ConvertKnowledgeGapResponse> {
  return apiRequest<ConvertKnowledgeGapResponse>(
    `/admin/query-analytics/gaps/${encodeURIComponent(gapId)}/convert`,
    { method: "POST", json: payload },
  );
}

export type DetectGapsRequest = {
  from_date?: string | null;
  to_date?: string | null;
  low_confidence_threshold?: number;
  min_occurrences?: number;
};

export type DetectGapsResponse = {
  detected: number;
  created: number;
  skipped_duplicates: number;
};

export async function detectKnowledgeGaps(
  payload: DetectGapsRequest = {},
): Promise<DetectGapsResponse> {
  return apiRequest<DetectGapsResponse>("/admin/query-analytics/gaps/detect", {
    method: "POST",
    json: payload,
  });
}

// ── Export ─────────────────────────────────────────────────────────────────────

export function buildQueryAnalyticsExportUrl(
  query: QueryAnalyticsSummaryQuery = {},
): string {
  const params = new URLSearchParams();
  if (query.from) params.set("from", query.from);
  if (query.to) params.set("to", query.to);
  const qs = params.toString();
  return `/api/admin/query-analytics/export${qs ? `?${qs}` : ""}`;
}

File: backend/app/domains/query_analytics/schemas/query_analytics.py
from __future__ import annotations

from datetime import date, datetime
from uuid import UUID

from pydantic import BaseModel, Field


# ── Date range helpers ─────────────────────────────────────────────────────────

class QueryAnalyticsDateRange(BaseModel):
    from_date: date
    to_date: date


# ── Trend data ─────────────────────────────────────────────────────────────────

class QueryTrendPoint(BaseModel):
    date: date
    total_queries: int
    unanswered: int
    low_confidence: int
    negative_feedback: int
    avg_confidence: float | None


class QueryTrendsResponse(BaseModel):
    organization_id: str
    range: QueryAnalyticsDateRange
    generated_at: datetime
    points: list[QueryTrendPoint]


# ── Dashboard summary ──────────────────────────────────────────────────────────

class FeedbackCategoryCount(BaseModel):
    category: str
    count: int


class QueryAnalyticsSummaryResponse(BaseModel):
    organization_id: str
    range: QueryAnalyticsDateRange
    generated_at: datetime
    enabled: bool
    disabled_reason: str | None = None

    total_queries: int
    answered_queries: int
    unanswered_queries: int
    low_confidence_queries: int
    negative_feedback_count: int

    unanswered_rate: float | None
    avg_confidence: float | None
    negative_feedback_rate: float | None

    top_feedback_categories: list[FeedbackCategoryCount]
    top_feedback_reasons: list[FeedbackCategoryCount]


# ── Knowledge gaps ─────────────────────────────────────────────────────────────

class KnowledgeGapResponse(BaseModel):
    gap_id: str
    organization_id: str
    gap_type: str
    topic_label: str
    description: str | None
    gap_source: str
    occurrence_count: int
    avg_confidence: float | None
    example_query: str | None
    status: str
    remediation_json: dict | None
    collection_id: str | None
    linked_document_id: str | None
    linked_eval_question_id: str | None
    converted_to: str | None
    converted_at: datetime | None
    reviewer_notes: str | None
    created_at: datetime
    updated_at: datetime


class KnowledgeGapListResponse(BaseModel):
    items: list[KnowledgeGapResponse]
    total: int


class CreateKnowledgeGapRequest(BaseModel):
    gap_type: str = Field(..., pattern="^(no_answer|low_confidence|bad_feedback|stale_citation|missing_source)$")
    topic_label: str = Field(..., min_length=1, max_length=256)
    description: str | None = Field(default=None, max_length=4096)
    occurrence_count: int = Field(default=1, ge=1)
    avg_confidence: float | None = Field(default=None, ge=0.0, le=1.0)
    example_query: str | None = Field(default=None, max_length=4096)
    collection_id: str | None = None
    gap_source: str = Field(default="admin", pattern="^(admin|low_confidence_analysis|feedback_analysis|no_answer_analysis)$")


class UpdateKnowledgeGapRequest(BaseModel):
    status: str | None = Field(default=None, pattern="^(open|in_review|resolved|dismissed)$")
    reviewer_notes: str | None = Field(default=None, max_length=4096)
    linked_document_id: str | None = None
    description: str | None = Field(default=None, max_length=4096)


class ConvertKnowledgeGapRequest(BaseModel):
    target: str = Field(..., pattern="^(eval_case|doc_request|review_task)$")
    evaluation_set_id: str | None = None
    notes: str | None = Field(default=None, max_length=4096)


class ConvertKnowledgeGapResponse(BaseModel):
    gap_id: str
    converted_to: str
    converted_at: datetime
    linked_eval_question_id: str | None = None


# ── Auto-detect gaps ───────────────────────────────────────────────────────────

class DetectGapsRequest(BaseModel):
    from_date: date | None = None
    to_date: date | None = None
    low_confidence_threshold: float = Field(default=0.5, ge=0.0, le=1.0)
    min_occurrences: int = Field(default=3, ge=1)


class DetectGapsResponse(BaseModel):
    detected: int
    created: int
    skipped_duplicates: int
